rotkehlchen/chain/evm/decoding/rainbow/decoder.py

```diff
@@ -0,0 +1,336 @@
+                in_event = event
+
+            # Find the spend event
+            if (event.event_type == HistoryEventType.SPEND and event.event_subtype == HistoryEventSubType.NONE and event.location_label == sender):  # noqa: E501
+                event.event_type = HistoryEventType.TRADE
+                event.event_subtype = HistoryEventSubType.SPEND
+                event.notes = f'Swap {event.amount} {event.asset.resolve_to_asset_with_symbol().symbol} in Rainbow Swap'  # noqa: E501
+                event.counterparty = CPT_RAINBOW_SWAPS
+                event.address = self.router_address
+                out_event = event
+
+        return in_event, out_event
+
+    def _create_fee_event(
+        self,
+        tx_hash: EVMTxHash,
+        sequence_index: int,
+        timestamp: int,
+        fee_raw: int,
+        fee_asset: Asset,
+        sender: ChecksumEvmAddress,
+    ) -> 'EvmEvent':
+        """Create a new fee event for rainbow swap
+           
+        Args:
+            tx_hash (EVMTxHash): transaction hash
+            sequence_index (int): event sequence index
+            timestamp (int): transaction timestamp
+            fee_raw (int): amount of fee charged in wei
+            fee_asset (Asset): fee asset
+            sender (ChecksumEvmAddress): sender address
+        """
+        fee_amount = asset_normalized_value(amount=fee_raw, asset=self._resolve_asset(fee_asset))
+        fee_event = self.base.make_event(
+            tx_hash=tx_hash,
+            sequence_index=sequence_index,
+            timestamp=timestamp,
+            event_type=HistoryEventType.SPEND,
+            event_subtype=HistoryEventSubType.FEE,
+            asset=fee_asset,
+            amount=fee_amount,
+            location_label=sender,
+            notes=f'Spend {fee_amount} {fee_asset.resolve_to_asset_with_symbol().symbol} as Rainbow Swap fees',  # noqa: E501
+            counterparty=CPT_RAINBOW_SWAPS,
+            address=self.router_address,
+        )
+        return fee_event
+
+    def _maybe_decode_swap_native_to_token(
+        self,
+        token: EvmToken | None,  # pylint: disable=unused-argument
+        tx_log: EvmTxReceiptLog,
+        transaction: EvmTransaction,
+        decoded_events: list['EvmEvent'],
+        action_items: list[ActionItem],   # pylint: disable=unused-argument
+        all_logs: list[EvmTxReceiptLog],
+    ) -> DecodingOutput:
+        """
+        Decodes a swap transaction from native token to another token.
+        Check if the first log is deposit and the last is transfer log
+        """
+        sender = transaction.from_address
+
+        # Check if is calling the rainbow router
+        if transaction.to_address != self.router_address:
+            return DEFAULT_DECODING_OUTPUT
+
+        # Check if the first log is wrap the native token
+        # and the last log is the transfer the token to the user
+        wrap_log = all_logs[0]
+        transfer_log = all_logs[-1]
```

Comment:
don't trust this because when using a safe or calling other contracts this can be off. For example the first event in safe transactions is the safe itself executing the call